### PR TITLE
[torch-dialect] fix aten.type_as op's folder

### PR DESCRIPTION
**Pull request by Vremold**
Previous folder for aten.type_as op didn't check whether the operands' dtype are all known. It might cause the aten.type_as op to be wrongly eliminated by `canonicalize` pass.

This patch canonicalize the aten.type_as op to aten.to_dtype op.

--- Edit ---
Change to use decomposition pass to convert aten.type_as op to prim.dtype + aten.to_dtype op.

### CODE REVIEW

File: lib/Dialect/Torch/IR/TorchOps.cpp [hunk 1]
```diff
@@ -716,14 +716,23 @@ OpFoldResult AtenRoundOp::fold(FoldAdaptor adaptor) {
...
+    Location loc = op.getLoc();
 
-  return nullptr;
+    Value targetDtype = rewriter.create<Torch::PrimDtypeOp>(loc, other);
+    Value nonBlocking = rewriter.create<Torch::ConstantBoolOp>(loc, false);
+    Value copy = rewriter.create<Torch::ConstantBoolOp>(loc, false);
+    Value memoryFormat = rewriter.create<Torch::ConstantNoneOp>(loc);
+    rewriter.replaceOpWithNewOp<Torch::AtenToDtypeOp>(
+        op, op.getType(), input, targetDtype, nonBlocking, copy, memoryFormat);
+    return success();
```
Comment (qingyunqu) [inline lib/Dialect/Torch/IR/TorchOps.cpp:734]: I think this pattern should be placed in DecomposeOps pass.

**Review by qingyunqu — CHANGES_REQUESTED** [review]

File: lib/Dialect/Torch/IR/TorchOps.cpp (hunk 1 above)
Comment (Vremold) [reply lib/Dialect/Torch/IR/TorchOps.cpp]: Done

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 2]
```diff
@@ -4754,6 +4777,7 @@ class DecomposeComplexOpsPass
     addPatternIfTargetOpIsIllegal<DecomposeAtenScalarTensor>(patterns);
     addPatternIfTargetOpIsIllegal<DecomposeAtenScatterValueOp>(patterns);
     addPatternIfTargetOpIsIllegal<DecomposeAtenSignOp>(patterns);
+    addPatternIfTargetOpIsIllegal<DecomposeAtenTypeAsOp>(patterns);
```
Comment (qingyunqu) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:4780]: you should also add this in:  https://github.com/llvm/torch-mlir/blob/38fb99df658faf46e1266a684b0900d183167371/lib/Dialect/Torch/Transforms/LowerToBackendContract.cpp#L364

**Review by qingyunqu — CHANGES_REQUESTED** [review]

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp (hunk 2 above)
Comment (Vremold) [reply lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:4780]: Done!

File: test/Dialect/Torch/decompose-complex-ops.mlir [hunk 3]
```diff
@@ -79,3 +79,27 @@ func.func @torch.aten.adaptive_avg_pool2d$unit_output_size(%arg0: !torch.vtensor
...
+// -----
+
+// CHECK-LABEL:  func.func @torch.aten.type_as$basic(
+// CHECK-SAME:                                %[[ARG_0:.*]]: !torch.tensor, %[[ARG_1:.*]]: !torch.tensor) -> !torch.tensor {
+// CHECK:          %[[FALSE:.*]] = torch.constant.bool false
+// CHECK:          %[[NONE:.*]] = torch.constant.none
+// CHECK:          %[[DTYPE:.*]] = torch.prim.dtype %[[ARG_1]] : !torch.tensor -> !torch.int
+// CHECK:          %[[VAR:.*]] = torch.aten.to.dtype %[[ARG_0]], %[[DTYPE]], %[[FALSE]], %[[FALSE]], %[[NONE]] : !torch.tensor, !torch.int, !torch.bool, !torch.bool, !torch.none -> !torch.tensor
+// CHECK:          return %[[VAR]] : !torch.tensor
+func.func @torch.aten.type_as$basic(%arg0: !torch.tensor, %arg1: !torch.tensor) -> !torch.tensor {
```
Comment (qingyunqu) [inline test/Dialect/Torch/decompose-complex-ops.mlir:92]: Also add an e2e test?

**Review by qingyunqu — CHANGES_REQUESTED** [review]

File: test/Dialect/Torch/decompose-complex-ops.mlir (hunk 3 above)
Comment (Vremold) [reply test/Dialect/Torch/decompose-complex-ops.mlir:92]: Done!

**Review by qingyunqu — APPROVED** [review]
LGTM

**Comment by Vremold** [comment]
Could you please also review the PR for me? @ramiro050

**Review by zhekunz2 — APPROVED** [review]
LGTM

**Review by ramiro050 — APPROVED** [review]

**Comment by Vremold** [comment]
Thanks!

**Pull request merged**